sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py

```diff
@@ -0,0 +1,170 @@
+            offset = offset or self._initial_event_position
+            consumer = self._eventhub_client.create_consumer(self._consumer_group_name, partition_id,
+                                                            EventPosition(str(offset)))
+            consumers.append(consumer)
+
+            partition_processor = self._partition_processor_factory(
+                eventhub_name=self._eventhub_name,
+                consumer_group_name=self._consumer_group_name,
+                partition_id=partition_id,
+                checkpoint_manager=CheckpointManager(partition_id, self._eventhub_name, self._consumer_group_name,
+                                                     self._instance_id, self._partition_manager)
+            )
+            loop = get_running_loop()
+            task = loop.create_task(
+                _receive(consumer, partition_processor, self._receive_timeout))
+            self._tasks.append(task)
+
+        await asyncio.gather(*self._tasks)
+        await self._partition_manager.close()
+        log.info("EventProcessor %r partition manager is closed", self._instance_id)
+        log.info("EventProcessor %r has stopped", self._instance_id)
+
+
+async def _receive(partition_consumer, partition_processor, receive_timeout):
+    try:
+        while True:
+            try:
+                events = await partition_consumer.receive(timeout=receive_timeout)
+            except asyncio.CancelledError:
+                await partition_processor.close(reason=CloseReason.SHUTDOWN)
+                log.info(
+                    "PartitionProcessor of EventProcessor instance %r of eventhub %r partition %r consumer group %r "
+                    "has been shutdown",
+                    partition_processor._checkpoint_manager._instance_id,
+                    partition_processor._eventhub_name,
+                    partition_processor._partition_id,
+                    partition_processor._consumer_group_name
+                )
+                break
+            except EventHubError as eh_err:
+                reason = CloseReason.LEASE_LOST if eh_err.error == "link:stolen" else CloseReason.EVENTHUB_EXCEPTION
+                log.info(
+                    "PartitionProcessor of EventProcessor instance %r of eventhub %r partition %r consumer group %r "
+                    "has met an exception receiving events. It's being closed. The exception is %r.",
+                    partition_processor._checkpoint_manager._instance_id,
+                    partition_processor._eventhub_name,
+                    partition_processor._partition_id,
+                    partition_processor._consumer_group_name,
+                    eh_err
+                )
+                await partition_processor.process_error(eh_err)
+                await partition_processor.close(reason=reason)
+                break
+            try:
+                await partition_processor.process_events(events)
```

Comment:
What happens if the task is cancelled while processing events? By default, it looks like any exceptions happening while running user code (including cancel) will be swallowed.

Reply:
Yes the exception shouldn't be swallowed. I've changed the code to handle it.
When cancel happens, process_events may finish half way. Next time when the partition processor is up again, it will start from the last checkpoint. Some messages may be processed more than once. 
Users can also intercept the CancelledError in process_events and do checkpoint before raise it again. They shouldn't swallow CancelledError. But this requires users to well understand the cancellation.

This may not be the best way. Let's see if there are better ways.